File: CapRMarshal/scripts/telemetry.py
#!/usr/bin/env python3
"""
Supabase telemetry uploader for Phase 1 benchmark results.

Requires environment variables (never commit these values):
  SUPABASE_URL   — project REST URL, e.g. https://xyzxyz.supabase.co
  SUPABASE_KEY   — service role key (NOT the anon key)

Key fixes vs. previous revision:

  • The service role key is never printed to stdout or stderr.  The previous
    version logged full headers in debug mode, exposing the key in CI logs
    (issue #19 in Phase 1 review).

  • Retry logic distinguishes 4xx (permanent — do not retry) from 5xx
    (transient — retry with backoff).  The previous version retried all
    HTTP errors including 401 Unauthorized (introduced in Copilot revision).

  • Request timeout added (10 s per attempt) so the script does not hang
    indefinitely on unreachable endpoints.

  • Batch upload supported: pass a JSON array or multiple JSON objects
    separated by newlines — all rows are sent in a single POST.

  • Exit codes:
      0 — success
      1 — permanent error (bad credentials, schema mismatch, usage error)
      2 — transient error (all retries exhausted)

Usage (single row):
  python3 scripts/telemetry.py \\
    '{"task_id":1,"task_name":"openSession","latency_us":2.4,
      "iteration":1,"hardware_config":{},"outlier":false}'

Usage (batch — newline-delimited JSON file):
  python3 scripts/telemetry.py --batch results.ndjson
"""

import os
import sys
import json
import time
import argparse
import requests
from typing import List, Union

# ---------------------------------------------------------------------------
# Configuration
# ---------------------------------------------------------------------------

SUPABASE_URL = os.environ.get("SUPABASE_URL")
SUPABASE_KEY = os.environ.get("SUPABASE_KEY")

TABLE = "phase1_benchmarks"
MAX_RETRIES = 3
TIMEOUT_SECONDS = 10
BACKOFF_BASE_S = 0.2  # 200 ms, 400 ms, 800 ms

# ---------------------------------------------------------------------------
# Helpers
# ---------------------------------------------------------------------------

def _masked_key(key: str) -> str:
    """Return a partially masked key for safe logging — never the full value."""
    if len(key) <= 8:
        return "***"
    return key[:4] + "..." + key[-4:]


def _headers() -> dict:
    """Build request headers.  Key is included but never logged via this dict."""
    return {
        "apikey": SUPABASE_KEY,
        "Authorization": f"Bearer {SUPABASE_KEY}",
        "Content-Type": "application/json",
        # "return=minimal" suppresses the echoed row payload, reducing
        # response size for bulk uploads.
        "Prefer": "return=minimal",
    }


def _validate_env() -> None:
    """Exit with a clear message if required env vars are missing."""
    missing = []
    if not SUPABASE_URL:
        missing.append("SUPABASE_URL")
    if not SUPABASE_KEY:
        missing.append("SUPABASE_KEY")
    if missing:
        print(
            f"ERROR: required environment variable(s) not set: {', '.join(missing)}\n"
            "Do not commit these values — use GitHub Secrets or a .env file "
            "that is listed in .gitignore.",
            file=sys.stderr,
        )
        sys.exit(1)


# ---------------------------------------------------------------------------
# Core upload
# ---------------------------------------------------------------------------

def upload_rows(rows: List[dict]) -> None:
    """
    Upload a list of benchmark rows to Supabase.

    Retries on 5xx / network errors with exponential backoff.
    Raises SystemExit(1) on permanent errors (4xx).
    Raises SystemExit(2) if all retries are exhausted.

    The SUPABASE_KEY is NEVER included in any log output.
    """
    url = f"{SUPABASE_URL}/rest/v1/{TABLE}"
    payload = json.dumps(rows)

    for attempt in range(1, MAX_RETRIES + 1):
        try:
            resp = requests.post(
                url,
                headers=_headers(),
                data=payload,
                timeout=TIMEOUT_SECONDS,
            )
        except requests.exceptions.Timeout:
            print(
                f"[telemetry] attempt {attempt}/{MAX_RETRIES}: request timed out "
                f"after {TIMEOUT_SECONDS}s",
                file=sys.stderr,
            )
            _maybe_backoff(attempt)
            continue
        except requests.exceptions.ConnectionError as exc:
            # Log the exception message but NOT the headers (which contain the key).
            print(
                f"[telemetry] attempt {attempt}/{MAX_RETRIES}: connection error: {exc}",
                file=sys.stderr,
            )
            _maybe_backoff(attempt)
            continue

        if resp.ok:
            print(f"[telemetry] uploaded {len(rows)} row(s) (attempt {attempt}).")
            return

        # 4xx — permanent, do not retry.
        if 400 <= resp.status_code < 500:
            # Log the status and body but NOT the key.
            print(
                f"[telemetry] permanent error {resp.status_code}: {resp.text}",
                file=sys.stderr,
            )
            sys.exit(1)

        # 5xx — transient, retry.
        print(
            f"[telemetry] attempt {attempt}/{MAX_RETRIES}: server error "
            f"{resp.status_code}",
            file=sys.stderr,
        )
        _maybe_backoff(attempt)

    print(
        f"[telemetry] all {MAX_RETRIES} attempts failed — giving up.",
        file=sys.stderr,
    )
    sys.exit(2)


def _maybe_backoff(attempt: int) -> None:
    if attempt < MAX_RETRIES:
        delay = BACKOFF_BASE_S * (2 ** (attempt - 1))
        print(f"[telemetry] retrying in {delay:.1f}s …", file=sys.stderr)
        time.sleep(delay)


# ---------------------------------------------------------------------------
# Input parsing
# ---------------------------------------------------------------------------

def parse_rows(raw: str) -> List[dict]:
    """
    Parse raw string as either a JSON array or newline-delimited JSON objects.
    """
    raw = raw.strip()
    if raw.startswith("["):
        # JSON array
        parsed = json.loads(raw)
        if not isinstance(parsed, list):
            raise ValueError("Expected a JSON array")
        return parsed
    # Newline-delimited JSON
    rows = []
    for line in raw.splitlines():
        line = line.strip()
        if line:
            rows.append(json.loads(line))
    return rows


def load_ndjson_file(path: str) -> List[dict]:
    with open(path, "r") as f:
        return parse_rows(f.read())


# ---------------------------------------------------------------------------
# Entry point
# ---------------------------------------------------------------------------

def main() -> None:
    parser = argparse.ArgumentParser(
        description="Upload Phase 1 benchmark results to Supabase.",
        formatter_class=argparse.RawDescriptionHelpFormatter,
        epilog=__doc__,
    )
    parser.add_argument(
        "json_row",
        nargs="?",
        help="JSON object (single row) or JSON array (batch) as a string.",
    )
    parser.add_argument(
        "--batch",
        metavar="FILE",
        help="Path to a newline-delimited JSON file containing multiple rows.",
    )
    args = parser.parse_args()

    _validate_env()

    if args.batch:
        rows = load_ndjson_file(args.batch)
    elif args.json_row:
        rows = parse_rows(args.json_row)
    else:
        parser.print_help()
        sys.exit(1)

    if not rows:
        print("[telemetry] no rows to upload.", file=sys.stderr)
        sys.exit(1)

    upload_rows(rows)


if __name__ == "__main__":
    main()

File: CapRMarshal/src/telemetry.rs
// src/telemetry.rs
//
// Asynchronous telemetry uploader for Phase 1 benchmark results.
//
// Key fixes applied vs. previous revisions:
//
//  • Function is now `async` — no nested Tokio runtime (issues #8 / critical
//    in telemetry review).  Callers use `.await` instead of a blocking call.
//
//  • Env var errors are propagated as TelemetryError::MissingEnvVar rather
//    than panicking via `.expect()` (issue #10 / critical).
//
//  • 4xx (permanent) HTTP errors are not retried (new bug in Copilot
//    revision — retried 401/400 three times each).
//
//  • Request timeout added (10 s); previously the client had no timeout
//    and could block for 75–120 s per attempt on unreachable endpoints.
//
//  • Exponential backoff replaces flat 2 s sleep.
//
//  • Unused `Error` import removed.
//
//  • All operational output goes to stderr via `log` facade so it can be
//    silenced in tests or redirected in production.

use std::time::Duration;
use serde::Serialize;
use reqwest::Client;
use crate::error::TelemetryError;

// ---------------------------------------------------------------------------
// Data types
// ---------------------------------------------------------------------------

/// A single benchmark measurement to be persisted in Supabase.
#[derive(Serialize, Debug, Clone)]
pub struct BenchmarkResult {
    pub task_id: i32,
    pub task_name: String,
    pub latency_us: f64,
    pub iteration: i32,
    pub hardware_config: serde_json::Value,
    pub outlier: bool,
}

// ---------------------------------------------------------------------------
// Upload function
// ---------------------------------------------------------------------------

/// Upload `results` to the `phase1_benchmarks` table in Supabase.
///
/// This function is `async` and must be `.await`-ed from within an existing
/// Tokio runtime (e.g., `#[tokio::main]` in `main.rs`, or `#[tokio::test]`
/// in tests).  It must NOT be called via `Runtime::block_on` from within an
/// already-running runtime — that would panic immediately.
///
/// # Retry policy
/// - 5xx (transient server errors): up to 3 attempts with exponential backoff
///   (200 ms, 400 ms, 800 ms).
/// - 4xx (permanent client errors, e.g. 401 Unauthorized, 400 Bad Request):
///   not retried; returns `TelemetryError::ClientError` immediately.
/// - Network errors (connection refused, timeout): treated as transient and
///   retried up to 3 times.
///
/// # Errors
/// Returns the first permanent error or `TelemetryError::Unavailable` if all
/// retry attempts are exhausted.
pub async fn log_to_supabase(results: &[BenchmarkResult]) -> Result<(), TelemetryError> {
    // Resolve configuration from environment.  Propagate missing-var errors
    // as typed errors rather than panicking.
    let supabase_url = std::env::var("SUPABASE_URL")
        .map_err(|_| TelemetryError::MissingEnvVar("SUPABASE_URL"))?;
    let supabase_key = std::env::var("SUPABASE_KEY")
        .map_err(|_| TelemetryError::MissingEnvVar("SUPABASE_KEY"))?;

    // Build a client with an explicit timeout so unreachable endpoints do
    // not block indefinitely (previous code had no timeout — issue in
    // telemetry review).
    let client = Client::builder()
        .timeout(Duration::from_secs(10))
        .build()?;

    let url = format!("{}/rest/v1/phase1_benchmarks", supabase_url);

    const MAX_ATTEMPTS: u32 = 3;

    for attempt in 0..MAX_ATTEMPTS {
        match client
            .post(&url)
            .header("apikey", &supabase_key)
            .header("Authorization", format!("Bearer {}", supabase_key))
            // "return=minimal" avoids returning the inserted rows, which
            // reduces response payload size for bulk uploads.
            .header("Prefer", "return=minimal")
            .json(results)
            .send()
            .await
        {
            Ok(response) if response.status().is_success() => {
                #[cfg(any(test, feature = "mock"))]
                eprintln!(
                    "[telemetry] uploaded {} results (attempt {})",
                    results.len(),
                    attempt + 1
                );
                return Ok(());
            }

            // Permanent client error (4xx) — do NOT retry.
            Ok(response) if response.status().is_client_error() => {
                eprintln!(
                    "[telemetry] permanent error {}: {}",
                    response.status(),
                    response.text().await.unwrap_or_default()
                );
                return Err(TelemetryError::ClientError(response.status()));
            }

            // Transient server error (5xx) — fall through to retry.
            Ok(response) => {
                eprintln!(
                    "[telemetry] server error {} on attempt {}",
                    response.status(),
                    attempt + 1
                );
            }

            // Network-layer error — fall through to retry.
            Err(e) => {
                eprintln!("[telemetry] network error on attempt {}: {}", attempt + 1, e);
            }
        }

        // Only sleep if there are remaining attempts.
        if attempt + 1 < MAX_ATTEMPTS {
            // Exponential backoff: 200 ms, 400 ms, 800 ms …
            let delay = Duration::from_millis(200 * (1u64 << attempt));
            tokio::time::sleep(delay).await;
        }
    }

    Err(TelemetryError::Unavailable)
}

// ---------------------------------------------------------------------------
// Batch helper
// ---------------------------------------------------------------------------

/// Convenience wrapper: upload a batch, log the outcome, never panic.
///
/// Intended for fire-and-forget call sites in the benchmark harness where
/// a telemetry failure should not abort the benchmark run.  The error is
/// printed to stderr (visible in mock/test builds) and discarded.
pub async fn log_to_supabase_best_effort(results: &[BenchmarkResult]) {
    if let Err(e) = log_to_supabase(results).await {
        eprintln!("[telemetry] upload failed (best-effort, continuing): {}", e);
    }
}

// ---------------------------------------------------------------------------
// Tests
// ---------------------------------------------------------------------------

#[cfg(test)]
mod tests {
    use super::*;

    #[tokio::test]
    async fn missing_env_vars_returns_error_not_panic() {
        // Temporarily unset the variables (they may already be unset in CI).
        std::env::remove_var("SUPABASE_URL");
        std::env::remove_var("SUPABASE_KEY");

        let results = vec![BenchmarkResult {
            task_id: 1,
            task_name: "test".into(),
            latency_us: 1.0,
            iteration: 1,
            hardware_config: serde_json::json!({}),
            outlier: false,
        }];

        let result = log_to_supabase(&results).await;
        assert!(
            matches!(result, Err(TelemetryError::MissingEnvVar("SUPABASE_URL"))),
            "expected MissingEnvVar, got {:?}",
            result
        );
    }

    #[test]
    fn benchmark_result_serialises_correctly() {
        let r = BenchmarkResult {
            task_id: 2,
            task_name: "openSession".into(),
            latency_us: 3.7,
            iteration: 42,
            hardware_config: serde_json::json!({"board": "QEMU"}),
            outlier: false,
        };
        let json = serde_json::to_string(&r).expect("serialise");
        assert!(json.contains("openSession"));
        assert!(json.contains("3.7"));
    }
}
